File: PlantHereNodeJs/Core/PlantHere.Aplication/Services/BasketService.js
// Interface - Service

const Interface = require("es6-interface");
const { IBasketService } = require('../../../Core/PlantHere.Aplication/Interfaces/Services/IBasketService')
const { PaymentService } = require('../../../Infratructure/PlantHere.Infratructure/PaymentService')

// RequestResponseModels

const { CreateBasketCommandResult } = require("../../../Core/PlantHere.Aplication/RequestResponseModels/Basket/Commands/CreateBasket/CreateBasketCommandResult")
const { CreateBasketCommand } = require("../../../Core/PlantHere.Aplication/RequestResponseModels/Basket/Commands/CreateBasket/CreateBasketCommand")

const { GetBasketByUserIdQuery } = require("../../../Core/PlantHere.Aplication/RequestResponseModels/Basket/Queries/GetBasketByUserId/GetBasketByUserIdQuery")
const { GetBasketByUserIdQueryResult } = require("../../../Core/PlantHere.Aplication/RequestResponseModels/Basket/Queries/GetBasketByUserId/GetBasketByUserIdQueryResult")

const { CreateBasketItemCommand } = require("../../../Core/PlantHere.Aplication/RequestResponseModels/BasketItem/Commands/CreateBasketItem/CreateBasketItemCommand")
const { CreateBasketItemCommandResult } = require("../../../Core/PlantHere.Aplication/RequestResponseModels/BasketItem/Commands/CreateBasketItem/CreateBasketItemCommandResult")

const { DeleteBasketItemCommand } = require("../../../Core/PlantHere.Aplication/RequestResponseModels/BasketItem/Commands/DeleteBasketItem/DeleteBasketItemCommand")
const { DeleteBasketItemCommandResult } = require("../../../Core/PlantHere.Aplication/RequestResponseModels/BasketItem/Commands/DeleteBasketItem/DeleteBasketItemCommandResult")

const { UpdateBasketItemCommand } = require("../../../Core/PlantHere.Aplication/RequestResponseModels/BasketItem/Commands/UpdateBasketItem/UpdateBasketItemCommand")
const { UpdateBasketItemCommandResult } = require("../../../Core/PlantHere.Aplication/RequestResponseModels/BasketItem/Commands/UpdateBasketItem/UpdateBasketItemCommandResult")

const { BuyBasketCommand } = require("../../../Core/PlantHere.Aplication/RequestResponseModels/Basket/Commands/BuyBasket/BuyBasketCommand")
const { BuyBasketCommandResult } = require("../../../Core/PlantHere.Aplication/RequestResponseModels/Basket/Commands/BuyBasket/BuyBasketComamndResult")


// Mapper
const { Mapper } = require('../../../Core/PlantHere.Aplication/Mapping/BasketProfile');

class BasketService extends Interface(IBasketService)
{
    constructor(repository) {
        super()
        this.repository = repository;
        this.paymentService = new PaymentService()
    }

    async createBasket(req) {
        return new CreateBasketCommandResult(await this.repository.createBasket(new CreateBasketCommand(req.userId)))
    }

    async getBasketByUserId(req) {

        const result = await this.repository.getBasketByUserId(new GetBasketByUserIdQuery(req.userId))
        return Mapper(result, GetBasketByUserIdQueryResult)
    }

    async buyBasket(req) {
        await this.repository.buyBasket(new BuyBasketCommand(req.userId, req.body.address))
    }

    async createBasketItem(req) {
        return new CreateBasketItemCommandResult(await this.repository.createBasketItem(new CreateBasketItemCommand(req.body.productId, req.userId, req.body.productName, req.body.price, req.body.discountedPrice)))
    }

    async deleteBasketItem(req) {
        return new DeleteBasketItemCommandResult(await this.repository.deleteBasketItem(new DeleteBasketItemCommand(req.userId, req.body.productId)))
    }

    async updateBasketItem(req) {
        return new UpdateBasketItemCommandResult(await this.repository.updateBasketItem(new UpdateBasketItemCommand(req.body.productId, req.userId, req.body.count)))
    }
}

module.exports = { BasketService }

File: PlantHere/PlantHere.Persistence/Services/BasketService.cs


using PlantHere.Application.CQRS.Basket.Commands.BuyBasket;
using PlantHere.Application.CQRS.Basket.Queries.GetBasketByUserId;
using PlantHere.Application.CQRS.BasketItem.Commands.CreateBasketItem;
using PlantHere.Application.CQRS.BasketItem.Commands.DeleteBasketItem;
using PlantHere.Application.CQRS.BasketItem.Commands.UpdateBasketItem;
using PlantHere.Domain.Aggregate.BasketAggregate.Entities;
using PlantHere.Domain.Aggregate.OrderAggregate.ValueObjects;

namespace PlantHere.Persistence.Services
{
    public class BasketService : Service<Basket>, IBasketService
    {
        private readonly IBasketRepository _basketReposity;
        private readonly IMapper _mapper;

        public BasketService(IRepository<Basket> repository, IUnitOfWork unitOfWork, IBasketRepository basketRepository, IMapper mapper) : base(repository, unitOfWork)
        {
            _basketReposity = basketRepository;
            _mapper = mapper;
        }

        public async Task<CustomResult<BuyBasketCommandResult>> BuyBasket(BuyBasketCommand buyBasketCommand)
        {
            var basket = await _basketReposity.GetBasketByUserId(buyBasketCommand.UserId);

            if (basket == null) throw new NotFoundException($"{typeof(Basket).Name}({buyBasketCommand.UserId}) Not Found");

            basket.BuyBasket(_mapper.Map<Address>(buyBasketCommand.Address), buyBasketCommand.Payment.CardTypeId, buyBasketCommand.Payment.CardNumber, buyBasketCommand.Payment.CardSecurityNumber, buyBasketCommand.Payment.CardHolderName);

            await _unitOfWork.CommitAsync();

            return CustomResult<BuyBasketCommandResult>.Success(200, new BuyBasketCommandResult());
        }

        public async Task<CustomResult<CreateBasketItemCommandResult>> CreateBasketItem(CreateBasketItemCommand createBasketItemCammand)
        {
            await _basketReposity.CreateBasketItem(createBasketItemCammand);
            await _unitOfWork.CommitAsync();
            return CustomResult<CreateBasketItemCommandResult>.Success(204, new CreateBasketItemCommandResult());
        }

        public async Task<CustomResult<DeleteBasketItemCommandResult>> DeleteBasketItem(DeleteBasketItemCommand deleteBasketItemCammand)
        {
            await _basketReposity.DeleteBasketItem(deleteBasketItemCammand);
            await _unitOfWork.CommitAsync();
            return CustomResult<DeleteBasketItemCommandResult>.Success(204, new DeleteBasketItemCommandResult());
        }

        public async Task<CustomResult<GetBasketByUserIdQueryResult>> GetBasketsByUserIdAsync(string userId)
        {
            var basket = await _basketReposity.GetBasketByUserId(userId);

            if (basket == null) throw new NotFoundException($"{typeof(Basket).Name}({userId}) Not Found");

            return CustomResult<GetBasketByUserIdQueryResult>.Success(200, _mapper.Map<GetBasketByUserIdQueryResult>(basket));
        }

        public async Task<bool> RemoveBasket(string userId)
        {
            var basket = await _basketReposity.GetBasketByUserId(userId);
            _basketReposity.Remove(basket);
            return true;
        }

        public async Task<CustomResult<UpdateBasketItemCommandResult>> UpdateBasketItem(UpdateBasketItemCommand updateBasketItemCammand)
        {
            await _basketReposity.UpdateBasketItem(updateBasketItemCammand);
            await _unitOfWork.CommitAsync();
            return CustomResult<UpdateBasketItemCommandResult>.Success(204, new UpdateBasketItemCommandResult());
        }
    }
}
